crypto/cms/cms_kari.c

```diff
@@ -342,18 +351,25 @@ static int cms_kari_set_originator_private_key(CMS_KeyAgreeRecipientInfo *kari,
 
 /* Initialise a kari based on passed certificate and key */
 
-int cms_RecipientInfo_kari_init(CMS_RecipientInfo *ri,  X509 *recip, EVP_PKEY *recipPubKey, X509 * originator, EVP_PKEY *originatorPrivKey, unsigned int flags)
+int cms_RecipientInfo_kari_init(CMS_RecipientInfo *ri,  X509 *recip,
+                                EVP_PKEY *recipPubKey, X509 *originator,
+                                EVP_PKEY *originatorPrivKey, unsigned int flags,
+                                const CMS_CTX *ctx)
 {
     CMS_KeyAgreeRecipientInfo *kari;
     CMS_RecipientEncryptedKey *rek = NULL;
 
     ri->d.kari = M_ASN1_new_of(CMS_KeyAgreeRecipientInfo);
-    if (!ri->d.kari)
+    if (ri->d.kari == NULL)
         return 0;
     ri->type = CMS_RECIPINFO_AGREE;
 
     kari = ri->d.kari;
     kari->version = 3;
+    kari->cms_ctx = ctx;
+
+    X509_set0_libctx(recip, ctx->libctx, ctx->propq);
+    X509_set0_libctx(originator, ctx->libctx, ctx->propq);
```

Comment:
We seem to do this a lot and I'm not convinced this is correct. Can't an X509 object have a lifetime beyond a single CMS message, and may be used in multiple CMS messages? So tying to set the libctx to the one used by the CMS_CTX somehow doesn't seem right. I would expect the libctx associated with the X509 to be set shortly after it is created, and then never modified again. If the X509 object is created as part of the CMS operation, then it seems reasonable to set its libctx (and do so immediately after its creation) - but if it is passed in through an external API then this makes less sense.

Reply:
fixed..